### PR TITLE
Retry

### PR DESCRIPTION
**Pull request by kyle-long**
It started out as implementing retries (which is also done) and
evolved into rewriting a lot of it including almost all of the
tests to be more functional so that refactors won't take so long
and so that the situations are more realistic

### CODE REVIEW

File: lib/response-handler.js [hunk 1]
```diff
@@ -18,53 +18,120 @@ module.exports = (bluebird, error, HttpLinkHeader, logger, ShelfError, URI) => {
...
+     *
+     * @param {superagent~Error} reason
+     * @return {shelfLib~ShelfError}
+     */
+    function createErrorForHttp(reason) {
+        var errorCode, errorCodeMap, mappedCode, statusCode;
+
+        errorCodeMap = {
+            404: error.NOT_FOUND,
+            500: error.INTERNAL_SERVER_ERROR
```
Comment (kyle-long) [inline lib/response-handler.js]: I suppose that seems fair enough. Question is, is this something the user would want to switch off of or do they just want to know it happened? It is part of the message, but it gets a code of UNKNOWN. Is there a reason the program would need to distinguish between 500 and 503?

Comment (quantumew) [inline lib/response-handler.js]: Yea good question. Perhaps we could give the user something that might be helpful to us if someone posted it in the Shelf Helpline. I will leave that to you though, whatever you think works best. I think doing it the same way we are doing a 500 makes sense error.GATEWAY_TIMEOUT and error.SERVICE_UNAVAILABLE but I am not bother to much either way.

File: lib/response-handler.js [hunk 2]
```diff
@@ -18,53 +18,120 @@ module.exports = (bluebird, error, HttpLinkHeader, logger, ShelfError, URI) => {
...
+     *
+     * @see {@link https://visionmedia.github.io/superagent/#error-handling}
+     *
+     * @param {superagent~Error} reason
+     * @return {ShelfError}
+     */
+    function createErrorForResponse(reason) {
+        var body, err;
+
+        logger.debug("In responseHandler.handleErrorResponse()");
```
Comment (kyle-long) [inline lib/response-handler.js]: Thats a strange line to comment on for that problem but yeah good catch.

Comment (quantumew) [inline lib/response-handler.js]: Haha I am lazy. I was like well that is a response so I will put it in the response handler

File: lib/response-handler.js [hunk 3]
```diff
@@ -95,7 +162,6 @@ module.exports = (bluebird, error, HttpLinkHeader, logger, ShelfError, URI) => {
 
         logger.debug("In responseHandler.resolveLink()");
         logger.debug(`Url param: ${resourceUrl}. Link param: ${link}`);
```
Comment (kyle-long) [inline lib/response-handler.js]: Yeah good catch. Wrote a test for relative links as well.

File: lib/artifact.js [hunk 4]
```diff
@@ -112,25 +64,17 @@ module.exports = (bluebird, fs, logger, request, requestOptions, requestPromise,
...
+
+            return this.shelfRequest.get(this.uri).then((response) => {
+                /*
+                 * In the future we should return a buffer instead of a
+                 * utf8 encoded string. I am only doing this here now
+                 * because the change is unrelated to the feature I
+                 * am working on right now and I don't want to try to
+                 * sneak in something that breaks the interface that bad.
+                 */
+                return response.body.toString("utf8");
```
Comment (kyle-long) [inline lib/artifact.js]: It all depends on how it is stored I suppose. I wouldn't expect it to be that much larger though I guess. I will make a note of it.

File: lib/artifact.js [hunk 5]
```diff
@@ -147,59 +91,13 @@ module.exports = (bluebird, fs, logger, request, requestOptions, requestPromise,
...
-                    })
-                    .on("error", (err) => {
-                        logger.debug(`In ${this.constructor.name}.downloadToFile() -> request.on("response").on("error)`);
-                        errorObj = responseHandler.createErrorForResponse(err);
-                        failed = true;
-
-                        reject(errorObj);
-                    });
-            });
+            return this.shelfRequest.downloadToStream(this.uri, downloadLocation);
```
Comment (kyle-long) [inline lib/artifact.js]: yeah no kidding heh. Worked out pretty nice

File: lib/index.js [hunk 6]
```diff
@@ -59,6 +60,10 @@ function shelfLibFactory(hostPrefix, libOptions) {
         libOptions.timeoutDuration = 300000;
     }
 
+    if (typeof libOptions.retries === "undefined") {
```
Comment (kyle-long) [inline lib/index.js]: We do care if retries are 0. I could specifically check for null or false if you'd like

File: lib/metadata.js [hunk 7]
```diff
@@ -1,17 +1,16 @@
...
          * @param {string} uri The URI for the artifact we want metadata for.
-         * @param {string} authToken The token used to authenticate access.
+         * @param {shelfLib~ShelfRequest} shelfRequest
          */
-        constructor(uri, authToken) {
-            this.authToken = authToken;
-            this.data = "";
+        constructor(uri, shelfRequest) {
+            this.shelfRequest = shelfRequest;
             this.uri = `${uri}/_meta`;
```
Comment (kyle-long) [inline lib/metadata.js]: made module

File: spec/helpers/create-test-lib.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+"use strict";
+var URI;
+
+URI = require("urijs");
+
+if (!jasmine.createTestLib) {
```
Comment (kyle-long) [inline spec/helpers/create-test-lib.js]: No reason apparently. I thought this was executed multiple times (for example once per test file). This is apparently not the case.

File: script/test [hunk 9]
```diff
@@ -11,7 +11,7 @@ PATH="$(pwd)/node_modules/.bin:$PATH"
 export PATH
 
 # Run the tests.
-istanbul cover jasmine
+istanbul cover jasmine -- "$@"
```
Comment (kyle-long) [inline script/test]: yeah heh

File: lib/response-handler.js (hunk 1 above)
Comment (kyle-long) [inline lib/response-handler.js]: As discussed I will just leave this be

File: lib/index.js (hunk 6 above)
Comment (quantumew) [inline lib/index.js]: Good point. It's all good

**Pull request merged**